include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,803 @@
+        : m_srcImage(std::move(srcImage)), m_dstImage(std::move(dstImage))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUImage> m_srcImage;
+    core::smart_refctd_ptr<const IGPUImage> m_dstImage;
+};
+
+class IGPUCommandPool::CCopyImageToBufferCmd : public IGPUCommandPool::IFixedSizeCommand<CCopyImageToBufferCmd>
+{
+public:
+    CCopyImageToBufferCmd(core::smart_refctd_ptr<const IGPUImage>&& srcImage, core::smart_refctd_ptr<const IGPUBuffer>&& dstBuffer)
+        : m_srcImage(std::move(srcImage)), m_dstBuffer(std::move(dstBuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUImage> m_srcImage;
+    core::smart_refctd_ptr<const IGPUBuffer> m_dstBuffer;
+};
+
+class IGPUCommandPool::CExecuteCommandsCmd : public IGPUCommandPool::IFixedSizeCommand<CExecuteCommandsCmd>
+{
+public:
+    CExecuteCommandsCmd(core::smart_refctd_dynamic_array<core::smart_refctd_ptr<const IGPUCommandBuffer>>&& commandBuffers) : m_commandBuffers(std::move(commandBuffers)) {}
+
+private:
+    core::smart_refctd_dynamic_array<core::smart_refctd_ptr<const IGPUCommandBuffer>> m_commandBuffers;
+};
+
+class IGPUCommandPool::CDispatchIndirectCmd : public IGPUCommandPool::IFixedSizeCommand<CDispatchIndirectCmd>
+{
+public:
+    CDispatchIndirectCmd(core::smart_refctd_ptr<const IGPUBuffer>&& buffer) : m_buffer(std::move(buffer)) {}
+
+private:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+};
+
+class IGPUCommandPool::CWaitEventsCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CWaitEventsCmd(const uint32_t bufferCount, const IGPUBuffer *const *const buffers, const uint32_t imageCount, const IGPUImage *const *const images, const uint32_t eventCount, IGPUEvent *const *const events)
+        : ICommand(calc_size(bufferCount, buffers, imageCount, images, eventCount, events)), m_resourceCount(bufferCount + imageCount + eventCount)
+    {
+        m_resources = new (this + sizeof(CWaitEventsCmd)) core::smart_refctd_ptr<const IReferenceCounted>[m_resourceCount];
+
+        uint32_t k = 0u;
+        for (auto i = 0; i < bufferCount; ++i)
+            m_resources[k++] = core::smart_refctd_ptr<const IReferenceCounted>(buffers[i]);
+
+        for (auto i = 0; i < imageCount; ++i)
+            m_resources[k++] = core::smart_refctd_ptr<const IReferenceCounted>(images[i]);
+
+        for (auto i = 0; i < eventCount; ++i)
+            m_resources[k++] = core::smart_refctd_ptr<const IReferenceCounted>(events[i]);
+    }
+
+    ~CWaitEventsCmd()
+    {
+        for (auto i = 0; i < m_resourceCount; ++i)
+            m_resources[i].~smart_refctd_ptr();
+    }
+
+    static uint32_t calc_size(const uint32_t bufferCount, const IGPUBuffer *const *const, const uint32_t imageCount, const IGPUImage *const *const, const uint32_t eventCount, IGPUEvent *const *const)
+    {
+        const uint32_t resourceCount = bufferCount + imageCount + eventCount;
+        return core::alignUp(sizeof(CWaitEventsCmd) + resourceCount * sizeof(void*), alignof(ICommand));
```

Comment:
for semantic clarity, use `alignof(CWaitEventsCmd)`

Reply:
Resolved.